agent/connect_auth.go

```diff
@@ -0,0 +1,146 @@
+package agent
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/consul/acl"
+	"github.com/hashicorp/consul/agent/cache"
+	cachetype "github.com/hashicorp/consul/agent/cache-types"
+	"github.com/hashicorp/consul/agent/connect"
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+// ConnectAuthorize implements the core authorization logic for Connect. It's in
+// a separate agent method here because we need to re-use this both in our own
+// HTTP API authz endpoint and in the gRPX xDS/ext_authz API for envoy.
+//
+// The ACL token and the auth request are provided and the auth decision (true
+// means authorised) and reason string are returned.
+//
+// If the request input is invalide the error returned will be a
+// BadRequestError, if the token doesn't grant necessary access then an
+// acl.ErrPermissionDenied error is returned, otherwise error indicates an
+// unexpected server failure. If access is denied, no error is returned but the
+// first return value is false.
+func (a *Agent) ConnectAuthorize(token string,
+	req *structs.ConnectAuthorizeRequest) (authz bool, reason string, m *cache.ResultMeta, err error) {
+	if req == nil {
+		err = BadRequestError{"Invalid request"}
+		return
+	}
+
+	// We need to have a target to check intentions
+	if req.Target == "" {
+		err = BadRequestError{"Target service must be specified"}
+		return
+	}
+
+	// Parse the certificate URI from the client ID
+	uri, err := connect.ParseCertURIFromString(req.ClientCertURI)
+	if err != nil {
+		err = BadRequestError{"ClientCertURI not a valid Connect identifier"}
```

Comment:
Is there value in exposing the details of the parsing failure when bubbling this up?

Reply:
I don't think so. Partly this is just copying the same functionality we already have for re-use so I didn't change it. This is generally a bug/user error they should be able to figure out if they know how our URIs are formed.